compiler/src/test/java/com/squareup/anvil/compiler/dagger/InjectConstructorFactoryGeneratorTest.kt

```diff
@@ -0,0 +1,608 @@
+package com.squareup.anvil.compiler.dagger
+
+import com.google.common.truth.Truth.assertThat
+import com.squareup.anvil.compiler.factoryClass
+import com.squareup.anvil.compiler.injectClass
+import com.squareup.anvil.compiler.isStatic
+import com.tschuchort.compiletesting.KotlinCompilation.Result
+import dagger.Lazy
+import dagger.internal.Factory
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.junit.runners.Parameterized
+import org.junit.runners.Parameterized.Parameters
+import java.io.File
+import javax.inject.Provider
+
+@Suppress("UNCHECKED_CAST")
+@RunWith(Parameterized::class)
+class InjectConstructorFactoryGeneratorTest(
+  private val useDagger: Boolean
+) {
+
+  companion object {
+    @Parameters(name = "Use Dagger: {0}")
+    @JvmStatic fun useDagger(): Collection<Any> {
+      return listOf(true, false)
```

Comment:
I run all tests for each code generator one time with Dagger and one time with Anvil. In both cases the generated code must pass the expectations. This way I make sure that the Anvil generates what Dagger expects, even though Anvil generates Kotlin code. Another safety net is that Dagger references the generated code when compiling components. If something doesn't match, then there is a compile error. When testing this feature in our codebase I found smaller edge cases at compile time, but at runtime there never was an issue.